### PR TITLE
update actions

### PR DESCRIPTION
**Pull request by rettinghaus**
This PR adds Python 3.12 to the test runs and updates the used GH actions. 
Supersedes #1205

### CODE REVIEW

File: .github/workflows/unit-test.yml [hunk 1]
```diff
@@ -23,18 +23,19 @@ jobs:
           - '3.9'
           - '3.10'
           - '3.11'
+          - '3.12'
         os:
           - ubuntu-22.04
           - ubuntu-20.04
-          # - macos-latest
+          - macos-latest
```
Comment (stweil) [inline .github/workflows/unit-test.yml:30]: Did you test the action with macos-latest?

**Review by stweil — APPROVED** [review]
The updates look good and are required to fix the CI annotations. Thank you!

**Review by kba — APPROVED** [review]
Thanks, we should use the latest versions of the actions.

**Comment by stweil** [comment]
Citing https://peps.python.org/pep-0632/:

> The distutils module [[1]](https://peps.python.org/pep-0632/#id10) has for a long time recommended using the setuptools package [[2]](https://peps.python.org/pep-0632/#id11) instead. Setuptools has recently integrated a complete copy of distutils and is no longer dependent on the standard library [[3]](https://peps.python.org/pep-0632/#id12). Pip has been silently replacing distutils with setuptools when installing packages for a long time already, and the distutils documentation has stated that it is being phased out since 2014 (or earlier). It is time to remove it from the standard library.

core still uses distutils. Therefore tests with Python 3.12 now fail.

**Comment by kba** [comment]
> Citing https://peps.python.org/pep-0632/:
> 
> > The distutils module [[1]](https://peps.python.org/pep-0632/#id10) has for a long time recommended using the setuptools package [[2]](https://peps.python.org/pep-0632/#id11) instead. Setuptools has recently integrated a complete copy of distutils and is no longer dependent on the standard library [[3]](https://peps.python.org/pep-0632/#id12). Pip has been silently replacing distutils with setuptools when installing packages for a long time already, and the distutils documentation has stated that it is being phased out since 2014 (or earlier). It is time to remove it from the standard library.
> 
> core still uses distutils. Therefore tests with Python 3.12 now fail.

I have a branch which removes distutils. The `find_executable` replacement is straightforward, I still need to test the `copy_tree` replacement.

**Comment by stweil** [comment]
@rettinghaus, why did you remove Python 3.12 in your latest commit? I think it is good to have a CI test with that version.

**Comment by stweil** [comment]
@kba, I suggest to merge this pull request (without the last commit which removed Python 3.12 from the test matrix).

**Comment by rettinghaus** [comment]
@stweil rebased and dropped last commit.

**Review by stweil — APPROVED** [review]

**Pull request merged**